motif/matchComp.go

```diff
@@ -0,0 +1,288 @@
+	var needNewKey bool = true
+	var currRefKey uint64
+	var couldGetNewKey, inKmerHash, couldScoreSequence bool
+	var currRefScore, currAltScore, currResidual, minResidual, minResidualAltScore float64
+	var refPos, lastRefPos, lastAlnPos, currAltStart, currAltEnd int = refStart, refStart, 0, 0, 0
+	var bitMask uint64 = uint64(math.Pow(2, float64(2*len(pm.Mat[0]))) - 1) // bitmask formula: B_n = 2^{2n} - 1
+	var currBed bed.Bed
+	revCompPm := ReverseComplement(pm)
+
+	// scan through reference sequence first
+	for alnPos := 0; alnPos < len(records[0].Seq); alnPos++ {
+		refPos = fasta.AlnPosToRefPosCounter(records[0], alnPos, lastRefPos, lastAlnPos)
+		//first we need a key for the alnPos. We can either make a new key or update our current key
+		if needNewKey {
+			currRefKey, alnPos, couldGetNewKey = getNewKey(records[0], alnPos, len(pm.Mat[0]))
+			refPos = fasta.AlnPosToRefPosCounter(records[0], alnPos, lastRefPos, lastAlnPos)
+			lastRefPos, lastAlnPos = refPos, alnPos
+			if !couldGetNewKey {
+				break //this means we've run out of windows on the reference sequence.
+			}
+			needNewKey = false
+		} else {
+			switch records[0].Seq[alnPos] {
+			case dna.N:
+				needNewKey = true
+				continue
+			case dna.Gap:
+				continue
+			case dna.A:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 0
+				currRefKey = currRefKey & bitMask
+			case dna.C:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 1
+				currRefKey = currRefKey & bitMask
+			case dna.G:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 2
+				currRefKey = currRefKey & bitMask
+			case dna.T:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 3
+				currRefKey = currRefKey & bitMask
+			default:
+				log.Fatalf("Unrecognized base.")
+			}
+		}
+
+		// now we check if the current key is a significant motif hit
+		if currRefScore, inKmerHash = kmerHash[currRefKey]; inKmerHash {
+			//fmt.Printf("CurrRefKey: %v. CurrRefScore: %v.\n", currRefKey, currRefScore)
+			minResidual = math.Inf(1)
+			minResidualAltScore = 0
+			for currAltStart = numbers.Max(alnPos-len(pm.Mat[0])-residualWindowSize+1, 0); currAltStart <= alnPos+residualWindowSize-len(pm.Mat[0])+1; currAltStart++ {
+				currAltScore, currAltEnd, couldScoreSequence = ScoreWindow(pm, records[1].Seq, currAltStart)
+				//fmt.Printf("AlnPos: %v. currAltStart: %v. CurrAltEnd: %v. CurrAltScore: %v.\n", alnPos, currAltStart, currAltEnd, currAltScore)
+				if !couldScoreSequence {
+					break
+				}
+				currResidual = math.Abs(currRefScore - currAltScore)
+				if currResidual < minResidual {
+					minResidual = currResidual
+					minResidualAltScore = currAltScore
+				}
+				if !enforceStrandMatch {
+					currAltScore, currAltEnd, couldScoreSequence = ScoreWindow(revCompPm, records[1].Seq, currAltStart)
```

Comment:
Is it necessary to do revCompPm, when in lines 73 and 77 you already do the scan sequence function for the reverse complement sequence?

Reply:
The situation for this one is that if our motif is CCC, and the ref is AAACCCAAA, and the alt is AAAGGGAAA, we should count no difference in motif, as the same position has the motif in the ref and alt, albeit the alt has the motif on the opposite strand. So we're doing the rapid 2bit kmer searching for the forward motif for the ref, but then checking both the forward and revcomp motif in the alt. However, this would still miss a GGG in the ref seq, so we'll need to rescan. Hope this clears it up, this is a funny nuance I'm happy to talk about

Reply:
Can you help me understand why missing the GGG (the revcomp motif) in the ref seq is a problem? I thought if you can find the CCC in the ref and the GGG in the alt that's enough, but I'm probably missing something

Reply:
revCompPm is to calculate revComp sequence score!